File: bindings/python/kevy/_cluster.py
"""Cluster-aware client (contract §3.15) + CRC16 slot routing (§7).

One connection per shard, CRC16-slot routed so single-key commands hit their
owner shard directly (no server ``-MOVED`` hop). Requires the server in
cluster mode; topology is discovered once at connect via ``CLUSTER SLOTS``
(16384 slots). The CRC16 reproduces Redis's ``key_hash_slot`` exactly
(including ``{hashtag}`` extraction) so routing agrees with the server.
"""

from __future__ import annotations

from typing import List, Optional, Tuple

from ._decode import Bytesish, score_of, to_bytes
from ._errors import ProtocolError, error_from_reply_text
from ._ops import to_ms
from ._reply import Reply, ReplyKind
from ._resp_conn import RespConn

_NUM_SLOTS = 16384
_CRC_POLY = 0x1021


def _make_crc_table() -> List[int]:
    table = []
    for i in range(256):
        crc = i << 8
        for _ in range(8):
            crc = ((crc << 1) ^ _CRC_POLY) & 0xFFFF if (crc & 0x8000) else (crc << 1) & 0xFFFF
        table.append(crc)
    return table


_CRC_TABLE = _make_crc_table()


def crc16(data: bytes) -> int:
    crc = 0
    for c in data:
        crc = ((crc << 8) ^ _CRC_TABLE[((crc >> 8) ^ c) & 0xFF]) & 0xFFFF
    return crc


def key_hash_slot(key: bytes) -> int:
    """Redis-cluster hash slot of key: crc16(hashtag(key)) & 16383."""
    return crc16(_hashtag(key)) & 0x3FFF


def _hashtag(key: bytes) -> bytes:
    open_i = key.find(b"{")
    if open_i < 0:
        return key
    close_i = key.find(b"}", open_i + 1)
    if close_i < 0:
        return key
    if close_i == open_i + 1:  # empty {} → hash whole key
        return key
    return key[open_i + 1 : close_i]


class ClusterClient:
    def __init__(self, shards: List[RespConn], slot_to_shard: List[int]):
        self._shards = shards
        self._slot_to_shard = slot_to_shard

    @classmethod
    def connect(cls, host: str, port: int) -> "ClusterClient":
        seed = RespConn.dial(host, port)
        try:
            reply = seed.request([b"CLUSTER", b"SLOTS"])
        finally:
            seed.close()
        ranges = _parse_cluster_slots(reply)
        nodes, table = _build_topology(ranges)
        shards: List[RespConn] = []
        try:
            for h, p in nodes:
                shards.append(RespConn.dial(h, p))
        except Exception:
            for s in shards:
                s.close()
            raise
        return cls(shards, table)

    def close(self) -> None:
        for s in self._shards:
            s.close()
        self._shards = []

    def __enter__(self) -> "ClusterClient":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    @property
    def shard_count(self) -> int:
        return len(self._shards)

    def _shard_for(self, key: bytes) -> RespConn:
        return self._shards[self._slot_to_shard[key_hash_slot(key)]]

    def request_keyed(self, key: Bytesish, *argv: Bytesish) -> Reply:
        return self._shard_for(to_bytes(key)).request([to_bytes(a) for a in argv])

    def request_unkeyed(self, *argv: Bytesish) -> Reply:
        return self._shards[0].request([to_bytes(a) for a in argv])

    def ping(self) -> None:
        r = self.request_unkeyed(b"PING")
        if r.kind is ReplyKind.SIMPLE and r.data in (b"PONG", b"OK"):
            return
        _raise_reply(r)

    def publish(self, channel: Bytesish, message: Bytesish) -> int:
        return _count(self.request_unkeyed(b"PUBLISH", channel, message))

    def set(self, key: Bytesish, value: Bytesish) -> None:
        _ok(self.request_keyed(key, b"SET", key, value))

    def set_with_ttl(self, key: Bytesish, value: Bytesish, ttl) -> None:
        _ok(self.request_keyed(key, b"SET", key, value, b"PX", str(to_ms(ttl)).encode()))

    def get(self, key: Bytesish) -> Optional[bytes]:
        r = self.request_keyed(key, b"GET", key)
        if r.kind is ReplyKind.BULK:
            return r.data
        if r.is_nil():
            return None
        _raise_reply(r)

    def incr(self, key: Bytesish) -> int:
        return _int(self.request_keyed(key, b"INCR", key))

    def incr_by(self, key: Bytesish, delta: int) -> int:
        return _int(self.request_keyed(key, b"INCRBY", key, str(int(delta)).encode()))

    def expire(self, key: Bytesish, ttl) -> bool:
        return _bool(self.request_keyed(key, b"PEXPIRE", key, str(to_ms(ttl)).encode()))

    def persist(self, key: Bytesish) -> bool:
        return _bool(self.request_keyed(key, b"PERSIST", key))

    def ttl_ms(self, key: Bytesish) -> int:
        return _int(self.request_keyed(key, b"PTTL", key))

    def delete(self, *keys: Bytesish) -> int:
        return self._per_key_sum(b"DEL", keys)

    def exists(self, *keys: Bytesish) -> int:
        return self._per_key_sum(b"EXISTS", keys)

    def _per_key_sum(self, verb: bytes, keys) -> int:
        total = 0
        for k in keys:
            n = _int(self.request_keyed(k, verb, k))
            if n > 0:
                total += n
        return total

    def dbsize(self) -> int:
        return _count(self.request_unkeyed(b"DBSIZE"))

    def flushall(self) -> None:
        _ok(self.request_unkeyed(b"FLUSHALL"))


# --- CLUSTER SLOTS parsing + topology -----------------------------------


def _build_topology(ranges: List[Tuple[int, int, str, int]]):
    if not ranges:
        raise ProtocolError("CLUSTER SLOTS returned no ranges")
    nodes: List[Tuple[str, int]] = []
    table = [0] * _NUM_SLOTS
    for start, end, host, port in ranges:
        node = (host, port)
        if node in nodes:
            idx = nodes.index(node)
        else:
            nodes.append(node)
            idx = len(nodes) - 1
        for slot in range(start, end + 1):
            table[slot] = idx
    return nodes, table


def _parse_cluster_slots(reply: Reply) -> List[Tuple[int, int, str, int]]:
    def bad():
        return ProtocolError("malformed CLUSTER SLOTS reply")

    if reply.kind is not ReplyKind.ARRAY:
        raise bad()
    out: List[Tuple[int, int, str, int]] = []
    for row in reply.items:
        if row.kind is not ReplyKind.ARRAY or len(row.items) < 3:
            raise bad()
        start = _as_int(row.items[0])
        end = _as_int(row.items[1])
        node = row.items[2]
        if start is None or end is None or node.kind is not ReplyKind.ARRAY or len(node.items) < 2:
            raise bad()
        host = _as_str(node.items[0])
        port = _as_int(node.items[1])
        if host is None or port is None:
            raise bad()
        out.append((start, end, host, port))
    return out


def _as_int(r: Reply) -> Optional[int]:
    return r.integer if r.kind is ReplyKind.INT else None


def _as_str(r: Reply) -> Optional[str]:
    if r.kind in (ReplyKind.BULK, ReplyKind.SIMPLE):
        return r.text()
    return None


def _raise_reply(r: Reply):
    if r.is_error():
        raise error_from_reply_text(r.data)
    raise ProtocolError(f"unexpected reply variant: {r.shape()}")


def _int(r: Reply) -> int:
    if r.kind is ReplyKind.INT:
        return r.integer
    _raise_reply(r)


def _count(r: Reply) -> int:
    n = _int(r)
    if n < 0:
        raise ProtocolError(f"expected non-negative count, got {n}")
    return n


def _bool(r: Reply) -> bool:
    if r.kind is ReplyKind.INT:
        return r.integer == 1
    _raise_reply(r)


def _ok(r: Reply) -> None:
    if r.kind is ReplyKind.SIMPLE and r.data == b"OK":
        return
    _raise_reply(r)

File: bindings/cpp/src/cluster.cpp
#include "kevy/cluster.hpp"

#include "argv.hpp"
#include "kevy/crc16.hpp"
#include "kevy/errors.hpp"
#include "reply_util.hpp"
#include "resp_conn.hpp"

namespace kevy {

using detail::i2s;
using detail::raise_reply_error;
using detail::raise_unexpected;

namespace {

constexpr int kNumSlots = 16384;

struct SlotRange {
  uint16_t start, end;
  std::string host;
  uint16_t port;
};

bool as_int(const Reply& r, int64_t& out) {
  if (r.kind == ReplyKind::Int) { out = r.integer; return true; }
  return false;
}
bool as_str(const Reply& r, std::string& out) {
  if (r.kind == ReplyKind::Bulk || r.kind == ReplyKind::Simple) { out = r.bytes; return true; }
  return false;
}
bool in_u16(int64_t n) { return n >= 0 && n <= 65535; }

std::vector<SlotRange> parse_cluster_slots(const Reply& reply) {
  auto bad = []() -> void { throw ProtocolError("malformed CLUSTER SLOTS reply"); };
  if (reply.kind != ReplyKind::Array) bad();
  std::vector<SlotRange> out;
  for (const auto& row : reply.array) {
    if (row.kind != ReplyKind::Array || row.array.size() < 3) bad();
    int64_t start = 0, end = 0, port = 0;
    std::string host;
    const Reply& node = row.array[2];
    if (!as_int(row.array[0], start) || !as_int(row.array[1], end)) bad();
    if (node.kind != ReplyKind::Array || node.array.size() < 2) bad();
    if (!as_str(node.array[0], host) || !as_int(node.array[1], port)) bad();
    if (!in_u16(start) || !in_u16(end) || !in_u16(port)) bad();
    out.push_back(SlotRange{static_cast<uint16_t>(start), static_cast<uint16_t>(end), host,
                            static_cast<uint16_t>(port)});
  }
  return out;
}

// Count expected non-negative; classify error.
int64_t cluster_int(const Reply& r) {
  if (r.kind == ReplyKind::Int) return r.integer;
  if (r.kind == ReplyKind::Error) raise_reply_error(r);
  raise_unexpected(r);
}
int64_t cluster_count(const Reply& r) {
  int64_t n = cluster_int(r);
  if (n < 0) throw ProtocolError("expected non-negative count, got " + i2s(n));
  return n;
}
bool cluster_bool(const Reply& r) {
  if (r.kind == ReplyKind::Int) return r.integer == 1;
  if (r.kind == ReplyKind::Error) raise_reply_error(r);
  raise_unexpected(r);
}
void cluster_ok(const Reply& r) {
  if (r.kind == ReplyKind::Simple && r.str() == "OK") return;
  if (r.kind == ReplyKind::Error) raise_reply_error(r);
  raise_unexpected(r);
}

}  // namespace

ClusterClient ClusterClient::connect(const std::string& host, uint16_t port) {
  auto seed = detail::RespConn::dial(host, port);
  Reply reply = seed->request({"CLUSTER", "SLOTS"});
  seed->close();
  std::vector<SlotRange> ranges = parse_cluster_slots(reply);
  if (ranges.empty()) throw ProtocolError("CLUSTER SLOTS returned no ranges");

  ClusterClient cc;
  cc.slot_to_shard_.assign(kNumSlots, 0);
  std::vector<std::pair<std::string, uint16_t>> nodes;
  for (const auto& r : ranges) {
    int idx = -1;
    for (size_t i = 0; i < nodes.size(); i++)
      if (nodes[i].first == r.host && nodes[i].second == r.port) { idx = static_cast<int>(i); break; }
    if (idx < 0) {
      nodes.emplace_back(r.host, r.port);
      idx = static_cast<int>(nodes.size()) - 1;
    }
    for (int slot = r.start; slot <= r.end; slot++) cc.slot_to_shard_[slot] = static_cast<uint16_t>(idx);
  }
  for (const auto& [h, p] : nodes) cc.shards_.push_back(detail::RespConn::dial(h, p));
  return cc;
}

ClusterClient::~ClusterClient() { close(); }
ClusterClient::ClusterClient(ClusterClient&&) noexcept = default;
ClusterClient& ClusterClient::operator=(ClusterClient&&) noexcept = default;

void ClusterClient::close() { shards_.clear(); }

detail::RespConn* ClusterClient::shard_for(std::string_view key) {
  return shards_[slot_to_shard_[key_hash_slot(key)]].get();
}

Reply ClusterClient::request_keyed(std::string_view key, const std::vector<std::string_view>& argv) {
  return shard_for(key)->request(argv);
}
Reply ClusterClient::request_unkeyed(const std::vector<std::string_view>& argv) {
  return shards_[0]->request(argv);
}

void ClusterClient::ping() {
  Reply r = request_unkeyed({"PING"});
  if (r.kind == ReplyKind::Simple && (r.str() == "PONG" || r.str() == "OK")) return;
  if (r.kind == ReplyKind::Error) raise_reply_error(r);
  raise_unexpected(r);
}

int64_t ClusterClient::publish(std::string_view channel, std::string_view message) {
  return cluster_count(request_unkeyed({"PUBLISH", channel, message}));
}

void ClusterClient::set(std::string_view key, std::string_view value) {
  cluster_ok(request_keyed(key, {"SET", key, value}));
}
void ClusterClient::set_with_ttl(std::string_view key, std::string_view value, ClusterDuration ttl) {
  int64_t ms = ttl.count() < 0 ? 0 : ttl.count();
  cluster_ok(request_keyed(key, {"SET", key, value, "PX", i2s(ms)}));
}
std::optional<std::string> ClusterClient::get(std::string_view key) {
  Reply r = request_keyed(key, {"GET", key});
  if (r.kind == ReplyKind::Bulk) return std::move(r.bytes);
  if (r.is_nil()) return std::nullopt;
  if (r.kind == ReplyKind::Error) raise_reply_error(r);
  raise_unexpected(r);
}
int64_t ClusterClient::incr(std::string_view key) {
  return cluster_int(request_keyed(key, {"INCR", key}));
}
int64_t ClusterClient::incr_by(std::string_view key, int64_t delta) {
  return cluster_int(request_keyed(key, {"INCRBY", key, i2s(delta)}));
}
bool ClusterClient::expire(std::string_view key, ClusterDuration ttl) {
  int64_t ms = ttl.count() < 0 ? 0 : ttl.count();
  return cluster_bool(request_keyed(key, {"PEXPIRE", key, i2s(ms)}));
}
bool ClusterClient::persist(std::string_view key) {
  return cluster_bool(request_keyed(key, {"PERSIST", key}));
}
int64_t ClusterClient::ttl_ms(std::string_view key) {
  return cluster_int(request_keyed(key, {"PTTL", key}));
}

int64_t ClusterClient::per_key_sum(const char* verb, const std::vector<std::string_view>& keys) {
  int64_t total = 0;
  for (auto k : keys) {
    int64_t n = cluster_int(request_keyed(k, {verb, k}));
    if (n > 0) total += n;
  }
  return total;
}
int64_t ClusterClient::del(const std::vector<std::string_view>& keys) { return per_key_sum("DEL", keys); }
int64_t ClusterClient::exists(const std::vector<std::string_view>& keys) { return per_key_sum("EXISTS", keys); }
int64_t ClusterClient::dbsize() { return cluster_count(request_unkeyed({"DBSIZE"})); }
void ClusterClient::flushall() { cluster_ok(request_unkeyed({"FLUSHALL"})); }

}  // namespace kevy
